### DISCUSSION TITLE
Start mongock in parallel

### DISCUSSION THREAD

**Discussion by sfgvieira** [flamingock/mongock#54] [Q&A] [answered]
Hello,

I am running a spring-boot application that is required to start very fast (a few seconds), however if I enable mongock depending on the nature of the migration it can take longer than that. I was wondering if there is a way (or recommended way) of starting mongock in parallel (and execute the migration) in order to not block the application startup.

Thanks!

**Comment by dieppa** [comment, marked as answer]
Hi @sfgvieira ,

first you need to understand the implications of  _"starting mongock in parallel (and execute the migration) in order to not block the application startup"_

If you do that, you can end up with an application that doesn't fit the data, as it can happen that applications starts without the migration finishing successfully, or even if it finishes fine, during the time the migration is running the code is already up. You can also do it the other way around, run first the migration and then deploy the application, which can produce the opposite scenario, teh data is up to date but the code isn't. In any case, unless you are confident and sure this is what you want, **We don't recommend this approach**

If you are OK with the implications, there may be more options, but these are two that would do the job. One for each approach:

1. **Migrate first:** This means you run the migration first and then deploy deploy the application. You can do this by using the **Mongock cli**(This is a new feature in version5. It's released, but we haven't published the official documentation. If you want to know more, please send us an email to dev@cloudyrock.io). Basically as part of your deployment, you run the migration first with the CLI and once finished, deploy the app.

2. **Deploy first:** Mongock offers three flavours when comes to the runner: The standard MongockRunner with the method `execute()` and for Springboot, MongockpplicationRunner and the MongockInitializingBean. The last two are automatically handled by Springboot and although both differ in the way they are started, both are executed before the application starts. So out of the box, you don't have it, but...
You can create an ApplicationRunner bean, which takes the MongockRunner and execute it asyncrhonosly. This means you cannot use `@EnableMongock` as it will create a MongockRunner bean at startup time.

Something like this:
```java
public class MongockApplicationRunnerAsync implements ApplicationRunner {

  private final MongockRunner runner;

  public MongockApplicationRunnerAsync(MongockRunner runner) {
    this.runner = runner;
  }

  @Override
  public void run(ApplicationArguments args) throws Exception {
    //You would probably use CompletableFuture or any other fancy framework, but you get the idea ;)
    new Thread(runner::execute).start();
  }
}
```
Then in your configuration class:
```java
@Bean
public MongockApplicationRunnerAsync MongockRunnerAsync(/**params**/) {
     MongockRunner runner = MongockSpringboot.builder()
          //...builder setters
          .buildRunner();
     return new MongockApplicationRunnerAsync(runner);
}
```

I haven't tested it myself, but it should do the job 😉
